versioned/tiered/rocksdb/src/main/java/org/projectnessie/versioned/rocksdb/RocksBaseValue.java

```diff
@@ -164,6 +175,259 @@ protected static String conditionNotMatchedMessage(Function function) {
+        .map(this::updateWithFunction)
+        .filter(Objects::nonNull)
+        .sorted(new Comparator<ExpressionPath.PathSegment>() {
+          /* Sort the PathSegments as follows
+           * names before positions
+           * names by string order
+           * higher positions before lower positions
+           * path with child before path without child
+           * when equal, recurse in with the child segments
+           */
+          @Override
+          public int compare(ExpressionPath.PathSegment ps1, ExpressionPath.PathSegment ps2) {
+            if (ps1.isName() && ps2.isName()) {
+              if (ps1.asName().getName().equals(ps2.asName().getName())) {
+                if (!ps1.getChild().isPresent() && ps2.getChild().isPresent()) {
+                  return 1;
+                } else if (ps1.getChild().isPresent() && !ps2.getChild().isPresent()) {
+                  return -1;
+                } else if (!ps1.getChild().isPresent() && !ps2.getChild().isPresent()) {
+                  return 0;
+                } else {
+                  return compare(ps1.getChild().get(), ps2.getChild().get());
+                }
+              } else {
+                return ps1.asName().getName().compareTo(ps2.asName().getName());
+              }
+            } else if (ps1.isName()) {
+              return -1;
+            } else if (ps1.isPosition() && ps2.isPosition()) {
+              return ps2.asPosition().getPosition() - ps1.asPosition().getPosition();
+            } else {
+              return 1;
+            }
+          }
+        }).collect(Collectors.toList());
+
+    for (ExpressionPath.PathSegment removedPath : removedPaths) {
+      if (removedPath.getChild().isPresent()) {
+        remove(removedPath.asName().getName(), removedPath.getChild().get());
+      }
+    }
+  }
+
+  /**
+   * Filters through the list of UpdateClause objects to remove conflicting updates. The order of
+   * the updates is preserved. Conflicts are defined as follows:
+   * <ul>
+   *   <li>Remove -&gt; Remove conflicts if latter remove is same node or a child node of former remove</li>
+   *   <li>Remove -&gt; SetEquals conflicts if SetEquals is for the same node or a child node of Remove</li>
+   *   <li>Remove -&gt; SetEquals conflicts if Remove is for a child node of SetEquals</li>
+   *   <li>Remove -&gt; Append conflicts if Append is for the same node or a child node of Remove</li>
+   *   <li>SetEquals -&gt; Remove conflicts if Remove is for a child node of SetEquals</li>
+   *   <li>SetEquals -&gt; SetEquals conflicts if latter SetEquals is for child node of the former SetEquals</li>
+   *   <li>SetEquals -&gt; Append conflicts if Append is for the same node or a child node of SetEquals</li>
+   * </ul>
+   * @param updateClauses The list of UpdateClause objects to filter
+   * @return A safe list of UpdateFunction objects to apply
+   */
+  private List<UpdateFunction> removeUnecessaryFunctions(List<UpdateClause> updateClauses) {
```

Comment:
Couldn't this be achieved in a simpler manner by simply ordering in the following?
 - All SetEquals, maintaining order
 - All Removes, maintaining order
 - All Appends, maintaining order

Reply:
The logic here covers all cases. There are likely many assumptions that could be made allowing this code to be much simpler.

I'm not sure that I follow what you mean here. Let's you divided up the updates into three categories, SetEquals, Removes, and Appends, then sort each category so that they are in a safe order. How do you apply these? Would you apply all SetEquals, then all Removes and finally all Appends?

If you apply all of the SetEquals first, then the following problem happens.
Original UpdateFunctions list:
1. SetEquals - /foo[0]
2. Remove - /foo[0]
3. SetEquals - /foo

In this case, 2 must be applied before 3, or 2 is ignored. In this example, it is safe to apply the Updates in order.

Consider this case:
1. SetEquals - /foo
2. SetEquals - /foo[0]
3. Remove - /foo[0]

I would argue that 2 and 3 are referring to the original value for "foo", so it is not safe to apply these Updates in order. Sorting just the SetEquals will not help.

This are a lot of if statements and there must be a better way of doing this. At the very least, I could structure this better.